controllers/virtualmachinesnapshot/virtualmachinesnapshot_controller.go

```diff
@@ -158,3 +184,215 @@ func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VirtualMachineSnapshotContext)
 	ctx.Logger.Info("Successfully patched VirtualMachine's current snapshot reference", "vm.Name", vm.Name, "spec.currentSnapshot", vm.Spec.CurrentSnapshot.Name)
 	return nil
 }
+
+func (r *Reconciler) ReconcileDelete(ctx *pkgctx.VirtualMachineSnapshotContext) error {
+	ctx.Logger.Info("Reconciling VirtualMachineSnapshot deletion")
+	vmSnapshot := ctx.VirtualMachineSnapshot
+	if controllerutil.ContainsFinalizer(vmSnapshot, Finalizer) {
+		vm := &vmopv1.VirtualMachine{}
+		objKey := client.ObjectKey{Name: vmSnapshot.Spec.VMRef.Name, Namespace: vmSnapshot.Namespace}
+		if err := r.Get(ctx, objKey, vm); err != nil {
+			if apierrors.IsNotFound(err) {
+				ctx.Logger.Info("VirtualMachine not found, assuming the snapshot is deleted along with moVM, remove finalizer")
+				controllerutil.RemoveFinalizer(vmSnapshot, Finalizer)
+				return nil
+			}
+			ctx.Logger.Error(err, "failed to get VirtualMachine", "vm", objKey)
+			return err
+		}
+
+		ctx.VM = vm
+		vmNotFound, err := r.deleteSnapshot(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to delete snapshot: %w", err)
+		}
+		if vmNotFound {
+			ctx.Logger.Info("VirtualMachine not found, assuming the snapshot is deleted along with moVM, remove finalizer", "error", err)
+			controllerutil.RemoveFinalizer(vmSnapshot, Finalizer)
+			return nil
+		}
+
+		parent, err := r.updateParentSnapshot(ctx)
```

Comment:
Just wondering, this controller, and the VM controller can modify the snapshot resource's status at the same time.  Do you see any issues with that if the parent of a snapshot was deleted _while_ the snapshot status was being populated by the VM controller?

Reply:
Then the `updateParentSnapshot` will fail, then it will trigger the reconcile again, and next time the parent snapshot won't be found and it shouldn't reach this code, will exit earlier when finding the parent snapshot.

Reply:
Well, you will need to handle currentSnapshot, rootSnapshot here because it is possible that the snapshot that is being deleted _is_ the only snapshot (so it is current, and is the root).
My PR handles the steady state update of the status whereas your PR handles one time fixup of these properties related to snapshot deletion.  I think we can keep this.

Reply:
OK, let see if we could observe conflict scenarios in live env. Can't think of good one now.